cassiopeia/solver/distance_correction/_crispr_cas9_distance_correction_solver.py

```diff
@@ -0,0 +1,556 @@
+    s1: List[int],
+    s2: List[int],
+    missing_state_indicator: int,
+) -> float:
+    """
+    Corrected ternary Hamming distance.
+
+    The "corrected" distance is an estimate of true tree distance.
+
+    It is assumed that the tree is ultrametric of depth 1.
+
+    Args:
+        mutation_proportion: Mutation proportion parameter of the model.
+        collision_probability: Collision probability parameter of the model.
+        s1: Character states of the first leaf.
+        s2: Character states of the second leaf.
+        missing_state_indicator: Missing state indicator.
+
+    Returns:
+        The corrected distance.
+    """
+    expected_ternary_hamming_distance_given_height = partial(
+        _crispr_cas9_expected_ternary_hamming_distance,
+        mutation_proportion=mutation_proportion,
+        collision_probability=collision_probability,
+    )
+    observed_ternary_hamming_distance = _ternary_hamming_distance(
+        s1=s1, s2=s2, missing_state_indicator=missing_state_indicator
+    )
+    height = _inverse(
+        f=expected_ternary_hamming_distance_given_height,
+        y=observed_ternary_hamming_distance,
+        lower=0,
+        upper=1,
+    )
+    estimated_tree_distance = 2 * height
+    return estimated_tree_distance
+
+
+def crispr_cas9_default_mutation_proportion_estimator(
+    character_matrix: pd.DataFrame,
+) -> float:
+    """
+    Estimate mutation proportion as #alleles>0/#alleles>=0.
+
+    Assumes that the missing data indicator is negative, and that all mutated
+    states are positive, for simplicity.
+
+    Args:
+        character_matrix: The character matrix.
+    Returns:
+        The estimated mutation proportion.
+    """
+    num_positive = (character_matrix > 0).sum().sum()
+    num_non_negative = (character_matrix >= 0).sum().sum()
+    if num_non_negative == 0:
+        # Everything is missing, avoid division by zero.
+        mutation_proportion = 0.0
+    else:
+        mutation_proportion = num_positive / num_non_negative
+    return mutation_proportion
+
+
+def crispr_cas9_hardcoded_mutation_proportion_estimator(
+    character_matrix: pd.DataFrame,
+    mutation_proportion: float,
+) -> float:
+    """
+    Hardcode the mutation proportion.
+
+    Args:
+        character_matrix: The character matrix (which will be ignored).
+        mutation_proportion: The mutation proportion to hardcode.
+    Returns:
+        The hardcoded `mutation_proportion`.
+    """
+    return mutation_proportion
+
+
+def crispr_cas9_default_collision_probability_estimator(
+    character_matrix: pd.DataFrame,
+) -> float:
+    """
+    Estimate collision probability by estimating individual state probabilities.
+
+    The estimated collision probability is:
+    q = sum q_i^2
+    where q_i is the estimated collision probability for state i using
+    #alleles=i / #alleles>0 (i.e. the number of times state i appears
+    in the character matrix, divided by the total number of mutations).
```

Comment:
I see below that this is just the fallback in case that a user does not provide a collision probability, and that the user has the opportunity to define their own collision probability estimator, but I can't help but feel we can provide more support in case the state probabilities _are_ passed into the solver. Why don't we add an extra optional argument called `priors` that will allow us to compute the more traditional collision probability for the user? And only if `priors=None` will we perform this naive estimation?

Reply:
Currently, if the user knows the collision probability they can just inject it into the `CRISPRCas9DistanceCorrectionSolver` using the `crispr_cas9_hardcoded_collision_probability_estimator`, as follows:

```
collision_probability_computed_from_priors = sum([p ** 2 for p in priors])
solver = CRISPRCas9DistanceCorrectionSolver(
    collision_probability_estimator=partial(
        crispr_cas9_hardcoded_collision_probability_estimator,
        collision_probability=collision_probability_computed_from_priors,
    )
)
```

This is exactly how I'm doing it for the experiments on the KP data, and I think it keeps the implementation of `CRISPRCas9DistanceCorrectionSolver` straightforward and clear. I am not a fan of conditional logic, but if you are thinking of something like this (where we use the tree's priors if they exist), I am happy to add it:

```
class CRISPRCas9DistanceCorrectionSolver(CassiopeiaSolver):
    def __init__(
        ...,
        use_priors_to_compute_collision_probability: bool = True,
    ):
        ...
        self._use_priors_to_compute_collision_probability = use_priors_to_compute_collision_probability

    def solve(...):
        if self._use_priors_to_compute_collision_probability and tree.priors is not None:
            collision_probability = sum([p ** 2 for p in priors])
        else:
            collision_probability = collision_probability_estimator(...)
```

I guess it would do more handholding for the user, at the expense of readability. Another option is to make the `CollisionProbabilityEstimatorType` a function of the `CassiopeiaTree` instead of a function of the character matrix, and then default it to the estimator that uses the priors if they are available in the tree and otherwise the current default. In this case, I am not a fan of taking a complex `CassiopeiaTree` object as an input because it makes it confusing what parts of the tree are being used to do the computation. Indeed, the `CassiopeiaTree` essentially acts like a huge data clump, and therefore saying that a function takes in `CassiopeiaTree` doesn't clarify what is being used unless you carefully read the docstring (in this case, the tree.priors would be used). I generally prefer functions that depend on simpler objects. They also require less context to be tested and are more reusable.